cpp/include/cuml/fil/multireduction.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+// data [T]ype, reduction [R]adix
+template <int R, typename T>
+__device__ T multireduction(T* per_thread, const int set_size, int n_sets) {
+  // reduce per-thread margin summand into per-class complete margin
+  // (for each of the NITEMS rows)
+  T acc;  // accumulator
```

Comment:
I wonder if the following could be a simpler version:

```
// reduce (=sum up) elements from data in n_groups groups, with each group containing n_values elements
// using all threads from the current block; data contains n_groups * n_values elements
template <typename T, int R>
__device__ T multi_reduction(T* data, int n_groups, int n_values) {
  T acc = threadIdx.x < n_groups * n_values ? data[threadIdx.x] : T();
  while (n_values > 1) {
    // ceildiv(x, y) = (x + y - 1) / y, i.e. division with rounding upwards
    // n_targets is the number of values per group after the end of this iteration
    int n_targets = ceildiv(n_values, R);
    if (threadIdx.x < n_targets * n_groups) {
      #pragma unroll
      for (int i = 1; i < R; ++i) {
        int idx = threadIdx.x + i * n_targets * n_groups;
        if (idx < n_values * n_groups) acc += data[idx];
        data[threadIdx.x] = acc;
      }
    }
    n_values = n_targets;
    if (n_values > 1) __syncthreads();
  }
  return acc;
}
```

Reply:
it is certainly simpler, however, it is somehow 10% to 20% slower than the current one, and tuning R did not help. it is still faster than the binary version.
do you have an idea of where this could be optimized?